security/s2a/internal/crypter/aesgcm_test.go

```diff
@@ -0,0 +1,319 @@
+			plaintext:   nil,
+			ciphertext:  nil,
+			tag:         dehex("250327c674aaf477aef2675748cf6971"),
+			nonce:       dehex("3c819d9a9bed087615030b65"),
+			result:      validResult,
+			allocateDst: true,
+		},
+		{
+			key:         dehex("ca47248ac0b6f8372a97ac43508308ed"),
+			plaintext:   nil,
+			ciphertext:  nil,
+			tag:         dehex("60d20404af527d248d893ae495707d1a"),
+			nonce:       dehex("ffd2b598feabc9019262d2be"),
+			result:      validResult,
+			allocateDst: true,
+		},
+		{
+			key:         dehex("7fddb57453c241d03efbed3ac44e371c"),
+			plaintext:   dehex("d5de42b461646c255c87bd2962d3b9a2"),
+			ciphertext:  dehex("2ccda4a5415cb91e135c2a0f78c9b2fd"),
+			tag:         dehex("b36d1df9b9d5e596f83e8b7f52971cb3"),
+			nonce:       dehex("ee283a3fc75575e33efd4887"),
+			result:      validResult,
+			allocateDst: true,
+		},
+		{
+			key:         dehex("ab72c77b97cb5fe9a382d9fe81ffdbed"),
+			plaintext:   dehex("007c5e5b3e59df24a7c355584fc1518d"),
+			ciphertext:  dehex("0e1bde206a07a9c2c1b65300f8c64997"),
+			tag:         dehex("2b4401346697138c7a4891ee59867d0c"),
+			nonce:       dehex("54cc7dc2c37ec006bcc6d1da"),
+			result:      validResult,
+			allocateDst: true,
+		},
+	} {
+		client, server := getGCMCryptoPair(test.key, t)
+		testGCMEncryptionDecryption(client, server, &test, t)
+	}
+}
+
+func testGCMEncryptRoundtrip(client S2AAeadCrypter, server S2AAeadCrypter, t *testing.T) {
+	// Construct a dummy nonce.
+	nonce := make([]byte, NonceSize)
+
+	// Encrypt.
+	const plaintext = "This is plaintext."
+	var err error
+	buf := []byte(plaintext)
+	buf, err = client.Encrypt(buf[:0], buf, nonce, nil)
+	if err != nil {
+		t.Fatal("Encrypting with client-side context: unexpected error", err, "\n",
+			"Plaintext:", []byte(plaintext))
+	}
+
+	// Decrypt first message.
+	ciphertext := append([]byte(nil), buf...)
+	buf, err = server.Decrypt(buf[:0], buf, nonce, nil)
+	if err != nil || string(buf) != plaintext {
+		t.Fatal("Decrypting client-side ciphertext with a server-side context did not produce original content:\n",
+			"  Original plaintext:", []byte(plaintext), "\n",
+			"  Ciphertext:", ciphertext, "\n",
+			"  Decryption error:", err, "\n",
+			"  Decrypted plaintext:", buf)
+	}
+
+	// Decryption fails: replay attack.
+	if got, err := server.Decrypt(nil, buf, nonce, nil); err == nil {
+		t.Error("Decrypting client-side ciphertext with a client-side context unexpectedly succeeded; want unexpected counter error:\n",
+			"  Original plaintext:", []byte(plaintext), "\n",
+			"  Ciphertext:", buf, "\n",
+			"  Decrypted plaintext:", got)
+	}
+}
+
+// Test encrypt and decrypt on roundtrip messages for AES-GCM.
+func TestAESGCMEncryptRoundtrip(t *testing.T) {
+	for _, keySize := range []int{AesGcmKeySize128, AesGcmKeySize256} {
+		key := make([]byte, keySize)
+		client, server := getGCMCryptoPair(key, t)
+		testGCMEncryptRoundtrip(client, server, t)
+	}
+}
+
+// Test encrypt and decrypt using an invalid key size.
```

Comment:
Can you add a test where we 1. create the crypter, 2. update the key, 3. use the new key to encrypt/decrypt. And also a test where we create the crypter, then update the key but it fails.